File: csharp/17_threads_and_processes.cs
// ============================================================
// Topic:  Multithreading & Multiprocessing  (FOCUS: Thread, lock, Mutex, Process)
// Summary: OS threads, synchronization, and separate processes.
// Run:     dotnet script csharp/17_threads_and_processes.cs
//
// KEY DIFFERENCES:
//   C#:     System.Threading.Thread, lock, Mutex, System.Diagnostics.Process
//   Python: threading.Thread, Lock, multiprocessing.Process
//   TS/JS:  Worker threads (worker_threads) — single-threaded event loop by default
//   Rust:   std::thread, Mutex<T>, std::process::Command
//   Java:   Thread, synchronized, ProcessBuilder
// ============================================================

using System;
using System.Diagnostics;
using System.Threading;

// ----- 1. Thread — start, run concurrently, Join -----
var workerReady = new ManualResetEventSlim(false);   // signal    | Python: threading.Event() | Java: CountDownLatch
Thread worker = new Thread(() =>
{
    Console.WriteLine("worker");                     // worker    | Python: print in thread target | Rust: thread::spawn closure
    workerReady.Set();                               // unblock main after worker prints
});
worker.Start();                                      // start     | Python: t.start() | Java: t.start()
workerReady.Wait();                                  // wait      | ensures worker prints before main
Console.WriteLine("main");                           // main      | main thread continues
worker.Join();                                       // join      | Python: t.join() | Rust: handle.join()
Console.WriteLine("joined");                         // joined    | worker finished

// ----- 2. lock — protect shared counter (two threads × 1000) -----
int counter = 0;
object counterLock = new object();                   // lock obj  | Python: threading.Lock() | Java: synchronized(this)
void IncrementCounter(int times)
{
    for (int i = 0; i < times; i++)
    {
        lock (counterLock)                           // lock      | Python: with lock: | Rust: mutex.lock()
        {
            counter++;
        }
    }
}
Thread t1 = new Thread(() => IncrementCounter(1000));
Thread t2 = new Thread(() => IncrementCounter(1000));
t1.Start();
t2.Start();
t1.Join();
t2.Join();
Console.WriteLine($"counter: {counter}");            // counter: 2000 | without lock would race below 2000

// ----- 3. Mutex — alternative sync primitive (ordered A then B) -----
Mutex abMutex = new Mutex();                         // Mutex     | cross-process capable | Python: multiprocessing.Lock
Thread threadA = new Thread(() =>
{
    abMutex.WaitOne();                               // acquire   | Python: lock.acquire()
    Console.WriteLine("A");                          // A         | first labeled output
    abMutex.ReleaseMutex();                          // release   | Python: lock.release()
});
Thread threadB = new Thread(() =>
{
    abMutex.WaitOne();
    Console.WriteLine("B");                          // B         | second after A (serialized by mutex)
    abMutex.ReleaseMutex();
});
threadA.Start();                                     // run A     | join before B for deterministic order
threadA.Join();
threadB.Start();
threadB.Join();
abMutex.Dispose();

// ----- 4. Process — separate address space, spawn child -----
var psi = new ProcessStartInfo
{
    FileName = "cmd.exe",                            // shell     | Python: subprocess.run(["cmd", "/c", ...])
    Arguments = "/c echo child",                     // child cmd | prints "child" from child process
    UseShellExecute = false,
};
Process child = Process.Start(psi)!;                 // spawn     | Python: multiprocessing.Process | Rust: Command::new
child.WaitForExit();                                 // wait      | block until child exits
Console.WriteLine($"exit: {child.ExitCode}");        // exit: 0   | 0 = success | Python: p.exitcode
child.Dispose();

// Cross-language note:
//   Thread = shared memory; Process = isolated memory — use threads for CPU-bound shared data, processes for isolation.
//   lock is syntactic sugar for Monitor.Enter/Exit — faster than Mutex when only one process needs sync.
//   Mutex works across processes; lock does not. Prefer lock for in-process counters.
//   Process.Start spawns a new OS process — heavier than Thread.Start but avoids GIL-style contention (N/A in C#).
//   dotnet script runs top-level statements; a .csproj needs explicit Main or host for the same flow.

File: rust/17_threads_and_processes.rs
// ============================================================
// Topic:  Multithreading & Multiprocessing
// Summary: std::thread::spawn, join, Mutex, std::process::Command child process.
// Run:     rustc --edition 2021 rust/17_threads_and_processes.rs -o rust/17_threads_and_processes.exe && rust/17_threads_and_processes.exe
//
// KEY DIFFERENCES:
//   Python:  threading.Thread + join, multiprocessing.Process, Lock
//   TS/JS:   Worker threads (worker_threads) / Web Workers; child_process.spawn
//   Rust:    std::thread, Arc<Mutex<T>>, std::process::Command (no GIL — true OS threads)
//   Java:    Thread / ExecutorService; ProcessBuilder for child processes
//   C++:     std::thread, std::mutex; fork/exec or std::system for child processes
// ============================================================

use std::process::Command;
use std::sync::{Arc, Mutex};
use std::thread;

fn main() {
    // ----- 1. spawn + join — worker thread runs, then main continues -----
    let handle = thread::spawn(|| {
        println!("worker"); // Output: worker   | Python: threading.Thread(target=worker).start()
    });
    handle.join().unwrap();
    println!("main"); // Output: main
    println!("joined"); // Output: joined   | Python: t.join()

    // ----- 2. Mutex — shared counter across threads (Arc = shared ownership) -----
    let counter = Arc::new(Mutex::new(0));
    let mut handles = vec![];
    for _ in 0..2 {
        let counter = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..1000 {
                *counter.lock().unwrap() += 1;
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    println!("counter: {}", *counter.lock().unwrap()); // Output: counter: 2000   | Python: Lock around += 1
    // Cross-language: Arc<Mutex<T>> = Python threading.Lock + shared ref; Rust enforces Send.

    // ----- 3. Two threads — sequential join for deterministic A then B -----
    let h_a = thread::spawn(|| println!("A")); // Output: A
    h_a.join().unwrap();
    let h_b = thread::spawn(|| println!("B")); // Output: B
    h_b.join().unwrap();
    // Python: Thread(target=print_a).start(); same pattern with join order.

    // ----- 4. Child process — Command spawns a shell echo -----
    #[cfg(windows)]
    let mut child = Command::new("cmd")
        .args(["/c", "echo", "child"])
        .spawn()
        .expect("failed to spawn child");
    #[cfg(not(windows))]
    let mut child = Command::new("sh")
        .args(["-c", "echo child"])
        .spawn()
        .expect("failed to spawn child");
    // Output: child   (printed by child stdout, inherited by parent)
    let status = child.wait().expect("failed to wait on child");
    println!("exit: {}", status.code().unwrap_or(-1)); // Output: exit: 0   | Python: subprocess.run(...).returncode

    // Cross-language note:
    //   Rust threads are 1:1 OS threads (no GIL). Mutex prevents data races at compile + runtime.
    //   Command = Python subprocess / Java ProcessBuilder / C++ std::system or posix_spawn.
    //   For heavy CPU parallelism across cores, same process threads share memory via Arc<Mutex>.
    //   Separate processes (Command) get isolated memory — like Python multiprocessing.Process.
}